core/common/src/main/java/alluxio/heartbeat/HeartbeatScheduler.java

```diff
@@ -54,25 +58,34 @@ private HeartbeatScheduler() {} // to prevent initialization
   public static void addTimer(ScheduledTimer timer) {
     Preconditions.checkNotNull(timer);
     try (LockResource r = new LockResource(sLock)) {
+      Preconditions.checkState(!sTimers.containsKey(timer.getThreadName()),
+          "The timer for thread %s is already waiting to be scheduled", timer.getThreadName());
       sTimers.put(timer.getThreadName(), timer);
       sCondition.signalAll();
     }
   }
 
   /**
-   * @param timer a timer to remove from the scheduler
+   * @param name the name of a timer to remove from the scheduler
    */
-  public static synchronized void removeTimer(ScheduledTimer timer) {
-    Preconditions.checkNotNull(timer);
+  public static void removeTimer(String name) {
```

Comment:
`removeTimer(timer)` should only remove the exact timer passed to the method, not another timer of the same name. `removeTimer(name)` should remove any timer with the given name. I'll add a check to `removeTimer(timer)` to enforce its contracts.